janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/step/JanusGraphMultiQueryStep.java

```diff
@@ -34,50 +34,78 @@
-    private final String forStep;
-    private boolean cachedStartsAccessed = false;
+    /**
+     * All steps that use this step to fill their cache. For example, this could be the
+     * next JanusGraphVertexStep. If the next step is a MultiQuery compatible parent
+     * (such as union()), then all of its child traversals can use this cache. Thus,
+     * there can be more than one client step.
+     */
+    private List<MultiQueriable> clientSteps = new ArrayList<>();
+    private final boolean limitBatchSize;
+    private boolean initialized;
 
-    public JanusGraphMultiQueryStep(Step<Vertex,?> originalStep) {
-        super(originalStep.getTraversal());
-        this.forStep = originalStep.getClass().getSimpleName();
+    public JanusGraphMultiQueryStep(Traversal.Admin traversal, boolean limitBatchSize) {
+        super(traversal);
+        this.limitBatchSize = limitBatchSize;
+        this.initialized = false;
     }
 
-    @Override
-    protected Admin<Vertex> processNextStart() throws NoSuchElementException {
-        Admin<Vertex> start = this.starts.next();
-        if (!cachedStarts.contains(start))
-        {
-            if (cachedStartsAccessed) {
-                cachedStarts.clear();
-                cachedStartsAccessed = false;
+    public void attachClient(MultiQueriable mq) {
+        clientSteps.add(mq);
+    }
+
+    private void initialize() {
+        assert !initialized;
+        initialized = true;
+
+        if (!limitBatchSize && !clientSteps.isEmpty()) { // eagerly cache all starts instead of batching
+            if (!starts.hasNext()) {
+                throw FastNoSuchElementException.instance();
             }
-            final List<Traverser.Admin<Vertex>> newStarters = new ArrayList<>();
-            starts.forEachRemaining(s -> {
-                newStarters.add(s);
-                cachedStarts.add(s);
+            final List<Traverser.Admin<Element>> elements = new ArrayList<>();
+            starts.forEachRemaining(e -> {
+                elements.add(e);
+                if (e.get() instanceof Vertex) {
+                    clientSteps.forEach(client -> client.registerFutureVertexForPrefetching((Vertex) e.get()));
+                }
             });
-            starts.add(newStarters.iterator());
-            cachedStarts.add(start);
+            starts.add(elements.iterator());
         }
-        return start;
     }
 
-    public List<Traverser.Admin<Vertex>> getCachedStarts() {
-        cachedStartsAccessed = true;
-        return new ArrayList<>(cachedStarts);
+    @Override
+    protected Admin<Element> processNextStart() throws NoSuchElementException {
+        if (!initialized) {
+            initialize();
+        }
+        Admin<Element> start = this.starts.next();
+        if (start.get() instanceof Vertex) {
+            clientSteps.forEach(client -> client.registerFutureVertexForPrefetching((Vertex) start.get()));
```

Comment:
When `limitBatchSize` is false, is this still needed? Seems all vertices would have been registered in the `initialize` method already.

Reply:
It should still work if I add `limitBatchSize && [...]` to the `if` condition. I will test that.

Reply:
It does not work. I forgot about the special case of `repeat()` steps. A step within a `repeat()` child does not have access to its `starts` when initialized for the second time. Therefore, we still need to handle these cases at some point.